### ISSUE TITLE
Asynch job and autowired bean

### ISSUE THREAD

**Issue by christophe1234** [paypal/resteasy-spring-boot#61]
Hello, 
first thanks for the job done. 

I would like to use Reasteasy and spring boot in my project but the configuration was a little complicated and with your project evreything was simpler. I am using the 2.3.0-RELEASE version

I have a problem with  [Asynchronous Job Service](https://docs.jboss.org/resteasy/docs/3.1.1.Final/userguide/html/async_job_service.html) Basically if I am requesting `http://localhost:8080/sample-app/echo?asynch=true` the response is synchronous. 

1. I have added `server.context-parameters.resteasy.async.job.service.enabled=true` in application.properties, the response is HTTP 404.
2. I have added too `server.context-parameters.resteasy.resources=com.test.Echo` the response is what we are expecting 
_HTTP/1.1 202
X-Application-Context: application
Location: http://localhost:8080/sample-app/asynch/jobs/1489333004352--164874753
Transfer-Encoding: chunked
Date: Sun, 12 Mar 2017 15:36:44 GMT_

The problem is when requesting http://localhost:8080/sample-app/asynch/jobs/1489333004352--164874753 there is no answer. 

When I am executing the program in debug mode the autowired property `echoer` is null.

Do you have an idea of the problem? What should I do to configure correctly the project to have asynchronous job service?
Thank you

**Comment by fabiocarvalho777** [comment]
Hello Christopher,

I am definitely going to take a look at it soon and, if necessary, provide a fix, or end user documentation entry, for the next version to be released.

Thanks for reporting this issue.

**Comment by fabiocarvalho777** [comment]
Hello,

I was able to reproduce this issue.

I don't have a solution for it yet, but I can tell you already that item 2 would not solve it, since resources are registered in another way when using this starter, [as documented here](https://github.com/paypal/resteasy-spring-boot/blob/master/mds/USAGE.md#resteasy-configuration).

I also don't have free time to work on it right now, but this issue is definitely on the top of my list whenever I am free, and I plan to add a solution for it in the next starter release, which happens usually every other month or so. In the meantime if you find a solution yourself please feel free to send a PR. I would be glad to review and merge it.

Regards,
Fabio Carvalho.

**Comment by christophe1234** [comment]
Hi,
thanks. I am trying to understand how everything is working :-) 
I was thinking that the solution could be found by modifying the method ResteasySpringBootConfig.resteasyBootstrapListener. 

When async.job.service.enabled is set to true (I have even tried so set it programmatically: deployment.setAsyncJobServiceEnabled(true) but same result), if I have understood the dispatcher your are setting is erased in the code of org.jboss.resteasy.spi.ResteasyDeployment.

Here the code of org.jboss.resteasy.spi.ResteasyDeployment.java
if (asyncJobServiceEnabled)
      {
         AsynchronousDispatcher asyncDispatcher = new AsynchronousDispatcher(providerFactory);
         asyncDispatcher.setMaxCacheSize(asyncJobServiceMaxJobResults);
         asyncDispatcher.setMaxWaitMilliSeconds(asyncJobServiceMaxWait);
         asyncDispatcher.setThreadPoolSize(asyncJobServiceThreadPoolSize);
         asyncDispatcher.setBasePath(asyncJobServiceBasePath);
         asyncDispatcher.getUnwrappedExceptions().addAll(unwrappedExceptions);
         dispatcher = asyncDispatcher;
         asyncDispatcher.start();  <-- What is the effect of the start here?
      }
      else
      {
         // If dispatcher is NOT null, that means it has already been set
         // previously, so we don' want to do it again, otherwise the original
         // one will be replaced
         if(dispatcher == null) {
            SynchronousDispatcher dis = new SynchronousDispatcher(providerFactory);
            dis.getUnwrappedExceptions().addAll(unwrappedExceptions);
            dispatcher = dis;
         }
}
 registry = dispatcher.getRegistry(); <-- your dispatcher should be used here (if async job is disabled)

I don't know if I am pointing on the problem or if I am completly out!

Regards,
Christophe

**Comment by fabiocarvalho777** [comment]
Hello,

You made a good point, that makes sense to me, and actually now I remembering things. [This check for the dispatcher here](https://github.com/resteasy/Resteasy/blob/3.1.0.Final/resteasy-jaxrs/src/main/java/org/jboss/resteasy/spi/ResteasyDeployment.java#L123-L126), was actually written by me in the RESTEasy code because I notice that that was causing problems in RESTEasy Spring Boot starter (overwriting the dispatcher we pre-configured in the starter). However, I forgot to add the same check for async dispatchers!!

If that is really true, and if that is all it takes, the solution will need:

1. Sending a PR to RESTEasy addressing the same issue I addressed for the sync cases, but this time for async.
1. Pre-configure the async dispatcher on the starter, similar to what is already done for the sync one.

This all makes sense to me.
For item 1 above we can test that change overwriting that code before sending a PR. For item 2, it is just a matter of implementing that on the starter code.

I will do all of that some time soon when I have free time, but please feel free to try to do it yourself, and send a PR, if you have free time sooner than me.

Great catch Christophe! :-)

**Comment by christophe1234** [comment]
Hello,
when I read your answer :-) , it reassures me that the bug is easy to correct. So I tried. 

1- ResteasyDeployment.java
if (asyncJobServiceEnabled)
		{
			if(dispatcher == null) {
				AsynchronousDispatcher asyncDispatcher = new AsynchronousDispatcher(providerFactory);
				asyncDispatcher.setMaxCacheSize(asyncJobServiceMaxJobResults);
				asyncDispatcher.setMaxWaitMilliSeconds(asyncJobServiceMaxWait);
				asyncDispatcher.setThreadPoolSize(asyncJobServiceThreadPoolSize);
				asyncDispatcher.setBasePath(asyncJobServiceBasePath);
				asyncDispatcher.getUnwrappedExceptions().addAll(unwrappedExceptions);
				dispatcher = asyncDispatcher;
				asyncDispatcher.start();
			}else{	
/*
*Perhaps we need to check if threads are already started?
*And to check that the dispatcher is of type  AsynchronousDispatcher?
*/
				((AsynchronousDispatcher)dispatcher).start();
			}

2 - ResteasySpringBootConfig.java
if(deployment.isAsyncJobServiceEnabled()){
					AsynchronousDispatcher dispatcher = new AsynchronousDispatcher(resteasyProviderFactory,resourceMethodRegistry);
					dispatcher.getUnwrappedExceptions().addAll(deployment.getUnwrappedExceptions());
					deployment.setDispatcher(dispatcher);
				}else{
					SynchronousDispatcher dispatcher = new SynchronousDispatcher(resteasyProviderFactory, resourceMethodRegistry);
					dispatcher.getUnwrappedExceptions().addAll(deployment.getUnwrappedExceptions());
					deployment.setDispatcher(dispatcher);
				}

And yes it works!!! Both sync and asynch request are working. 
What do you want to do? I have never opened a PR. Or do you prefer to do it?

Have a nice week,
Christophe

**Comment by fabiocarvalho777** [comment]
Thanks Christophe!

Regarding item 1, I can do it soon (probably next week or so), or you can do it, if sooner than me. Your call. If you do it, you can send a PR for this starter for the changes at `ResteasySpringBootConfig.java` file. You will need to fork this repo, checkout develop branch, apply your changes, then send a PR. You can search in google how to fork, checkout and send PRs using Git and GitHub. Also, please follow [these guidelines here](https://github.com/paypal/resteasy-spring-boot/blob/develop/mds/CONTRIBUTING.md). Don't forget please to add, in the same PR that has the fix, an integration test to `resteasy-spring-boot-starter-test` project to expose and test this issue. That might be actually trickier than the fix itself.

Regarding item 2, the changes at RESTEasy itself, don't worry, I will do it.

Thanks!

**Comment by fabiocarvalho777** [comment]
Hello @christophe1234,

Just checking, have you started working on this? If not, let me know, I have some time this week then I would work on it.

Thanks.

**Comment by christophe1234** [comment]
Hello  @fabiocarvalho777,
sorry no time to work on it. I am doing tonight the point 1:  the update of ResteasySpringBootConfig.java with the PR.
Thank you

**Comment by christophe1234** [comment]
Hello,
I forked the project. But like you said the test are trickies.I will se tonight to commit a test but not sure to do it properly.

Christophe

**Comment by fabiocarvalho777** [comment]
Thanks for letting me know. If you prefer, you can just push the actual code change, I can work with the test. Just let me know.

**Comment by christophe1234** [comment]
Hi,
I pushed my code in the branch: https://github.com/christophe1234/resteasy-spring-boot/tree/develop
Christophe

**Comment by fabiocarvalho777** [comment]
Cool, thanks.

Now you need to send a Pull Request from your fork. You can do that by pressing the "New pull request" button.

But before that, notice that your commit diff has many lines changed that are not relevant. That is because you are using tabs instead of spaces. Please, revert that, use 4 spaces for indentation.

**Comment by fabiocarvalho777** [comment]
I am working on this issue. I have already [added the necessary changes](https://github.com/paypal/resteasy-spring-boot/commit/844c7f1b891fff8a9cf2b8b90cb4c207cde38415) to this starter project, to RESTEasy ([under this branch in my fork](https://github.com/fabiocarvalho777/Resteasy/tree/custom_async_dispatch)), and also [an integration test](https://github.com/paypal/resteasy-spring-boot/blob/issue_61/resteasy-spring-boot-starter-test/src/test/java/com/paypal/springboot/resteasy/AsyncJobIT.java). I placed all this code in branch [issue_61](https://github.com/paypal/resteasy-spring-boot/tree/issue_61).
A few tests are failing though. I will fix them and then get all this changes pushed finally.

**Comment by christophe1234** [comment]
Sorry to let you correct everything. I had a look at your modifications and tests. Impossible to launch projects, I think I have a problem of internet connexion. But for me, your updates seem alright.

**Comment by fabiocarvalho777** [comment]
It is ok Christophe.

This fix is incomplete though. It registers the async dispatcher, but it doesn't work as it should, it is working synchronously, instead of asynchronously. I will keep working on it to fix it soon. After everything works fine, then I will send the PR to RESTEasy.

**Comment by christophe1234** [comment]
Hi,
1- so first I tested your modifications by sending http requests with curl after launching the sample-app project (server.context-parameters.resteasy.async.job.service.enabled=true in application.properties) and everything is working for me.

2- Secondly I launched your tests in AsyncJobIT. I have got 2 failures: asyncRequestTest and fireAndForgetRequestTest.
The problem is on the "body(empty())" test. Is "Empty matcher" working only on collection? What do you think by replacing it by "isEmptyOrNullString()"? After tests are correct.

3 - I deleted the application.properties in the sample-app and recompilethe project. Tests are passing.

For me it is working asynchronusly. Coudl you give me more details about the tests you are doing? Perhaps I miss something in my tests.

Christophe

**Comment by fabiocarvalho777** [comment]
Hello,

1- Async requests are definitely not working. When you send an async request, [as described here](https://docs.jboss.org/resteasy/docs/3.1.1.Final/userguide/html/async_job_service.html), you are supposed to receive a 202 response, with no payload, and a `Location` header. None of that is happening. If you are receiving a response message with a content in the payload that means it is not working, it is just behaving as the request was sync, which is not the expected behavior.

2- Those two tests are expected to fail, as I mentioned in my comment on April 30th. They fail exactly because they were correctly written to assure the request were processed asynchronously, but they were not, as I say in item 1 above. As soon as item 1 is fixed, then those two tests should pass. So, nothing wrong with the tests, but with item 1.

3- It does not make sense to remove application.properties. They are configured exactly to support all test cases, as they are.

I hope to find some free time this week to conclude this work. I would say 50% is done with the changes I have pushed so far.

Fábio.

**Comment by fabiocarvalho777** [comment]
Hello Christophe, sorry this is taking time to get done. I will definitely go back to it and finish it soon. Right now I have two other issues going on in this project and I am working with them in parallel too, besides this one, so it's been a busy time.

**Comment by fabiocarvalho777** [comment]
Hello Christophe,

First of all, I apologize, you were right! My solution was correct and complete, all it needed was fixing those two tests, replacing the `empty` check by `isEmptyString`. The final solution is already available in branch [issue_61](https://github.com/paypal/resteasy-spring-boot/tree/issue_61).

Also, as you know, it needs a change in RESTEasy as well. I have already reported it to RESTEasy team, [here is the issue](https://issues.jboss.org/browse/RESTEASY-1655). I am going to provide a PR for it very soon. Ass soon as they merge it, I will be able to merge the solution for issue 61 to the starter develop branch.

Thanks!

**Comment by fabiocarvalho777** [comment]
I have just pushed a PR to RESTEasy, here it is: https://github.com/resteasy/Resteasy/pull/1167

**Comment by christophe1234** [comment]
Hello Fabio,
good news. I am happy that we are both agree that your fix is correct and working :-) 

When all packages will be released, I will intregate them in my project. 

Thank you for your work
Christophe

**Comment by fabiocarvalho777** [comment]
Another good news, RESTEasy has already merged our PR, and it will be released with the next RESTEasy version, 3.1.3.Final.
So, as soon as that is released, then I can release a RESTEasy starter version with this fix.

**Comment by ui4j** [comment]
RestEasy 3.1.3.Final is released.

https://issues.jboss.org/browse/RESTEASY-1655 (Closed)

**Comment by fabiocarvalho777** [comment]
This issue is finally done. The solution is available in `com.paypal.springboot:resteasy-spring-boot-starter-parent:2.3.2-SNAPSHOT`. I will release version `2.3.2-RELEASE` soon.

**Comment by fabiocarvalho777** [comment]
@christophe1234 Notice that version 2.3.2-RELEASE was released yesterday with this fix.
